Crash report: None.get, it does not like 'requires x'

So if I write 'requires x' the program crashes when verification fails. Requires 'x==true' works as expected.

It looks strange, in ProcessError (SilverBackend.scala:199) the reason.offendingNode is 'x==true' when I use "requires x==true".

But when I use 'requires x' it will give back '1==0' as offendingNode. So maybe the problem lies with silicon?

# Crash Message
```
None.get
  at scala.None$.get(Option.scala:627)
  at scala.None$.get(Option.scala:626)
  at viper.api.backend.SilverBackend.path(SilverBackend.scala:80)
  at viper.api.backend.SilverBackend.processError(SilverBackend.scala:199)
  at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:173)
  at viper.api.backend.silicon.Silicon.processError(Silicon.scala:34)
  at viper.api.backend.SilverBackend.$anonfun$submit$1(SilverBackend.scala:165)
  at viper.api.backend.SilverBackend.$anonfun$submit$1$adapted(SilverBackend.scala:165)
  at scala.collection.immutable.List.foreach(List.scala:333)
  at viper.api.backend.SilverBackend.submit(SilverBackend.scala:165)
  at viper.api.backend.SilverBackend.submit$(SilverBackend.scala:144)
  at viper.api.backend.silicon.Silicon.submit(Silicon.scala:34)
  ...
  at vct.main.Main$.main(Main.scala:51)
  at vct.main.Main.main(Main.scala)
```

# Version Information
* VerCors version `9999.9.9-SNAPSHOT`
* At commit dd96e8678302f29c8e2b620f93a68858d8b5b8b0 from branch `dev` (changes=true)

# Arguments
`/home/lars/data/vercors/examples/concepts/c/floats.c` `--dev-no-sat` `--no-infer-heap-context-into-frame` `--dev-total-timeout=3600` `--backend-file-base` `/home/lars/data/vercors/examples/concepts/c/floats`

# File Inputs
<details>
<summary>floats.c</summary>

```c
#include <stdbool.h>


/*@
 requires x;
 requires x == true;
*/
void f(bool x){
    return;
}


int test2 (){
  f(1==0);
  return 0;
}   
```
</details>


# Full Log
<details>

```
11:27:17.594 [INFO] Start: VerCors (at 11:27:17)
11:27:17.685 [WARN] Caching is enabled, but results will be discarded, since there were uncommitted changes at compilation time.
11:27:25.381 [INFO] Done: VerCors (at 11:27:25, duration: 00:00:07)
11:27:25.384 [ERROR] java.util.NoSuchElementException: None.get
	at scala.None$.get(Option.scala:627)
	at scala.None$.get(Option.scala:626)
	at viper.api.backend.SilverBackend.path(SilverBackend.scala:80)
	at viper.api.backend.SilverBackend.processError(SilverBackend.scala:199)
	at viper.api.backend.SilverBackend.processError$(SilverBackend.scala:173)
	at viper.api.backend.silicon.Silicon.processError(Silicon.scala:34)
	at viper.api.backend.SilverBackend.$anonfun$submit$1(SilverBackend.scala:165)
	at viper.api.backend.SilverBackend.$anonfun$submit$1$adapted(SilverBackend.scala:165)
	at scala.collection.immutable.List.foreach(List.scala:333)
	at viper.api.backend.SilverBackend.submit(SilverBackend.scala:165)
	at viper.api.backend.SilverBackend.submit$(SilverBackend.scala:144)
	at viper.api.backend.silicon.Silicon.submit(Silicon.scala:34)
	at vct.main.stages.SilverBackend.verify(Backend.scala:201)
	at vct.main.stages.Backend.$anonfun$run$6(Backend.scala:166)
	at vct.main.stages.Backend.$anonfun$cachedDefinitelyVerifiesOrElseUpdate$1(Backend.scala:106)
	at scala.Option.getOrElse(Option.scala:201)
	at vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate(Backend.scala:104)
	at vct.main.stages.Backend.cachedDefinitelyVerifiesOrElseUpdate$(Backend.scala:100)
	at vct.main.stages.SilverBackend.cachedDefinitelyVerifiesOrElseUpdate(Backend.scala:184)
	at vct.main.stages.Backend.$anonfun$run$5(Backend.scala:166)
	at vct.main.stages.Backend.$anonfun$run$5$adapted(Backend.scala:163)
	at hre.progress.Progress$.$anonfun$foreach$2(Progress.scala:25)
	at scala.runtime.java8.JFunction0$mcV$sp.apply(JFunction0$mcV$sp.scala:18)
	at hre.progress.task.AbstractTask.frame(AbstractTask.scala:166)
	at hre.progress.Progress$.$anonfun$foreach$1(Progress.scala:25)
	at hre.progress.Progress$.$anonfun$foreach$1$adapted(Progress.scala:24)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterator.foreach(Iterator.scala:1300)
	at hre.progress.Progress$.foreach(Progress.scala:24)
	at vct.main.stages.Backend.$anonfun$run$1(Backend.scala:163)
	at hre.progress.task.NameSequenceTask.scope(NameSequenceTask.scala:16)
	at hre.progress.Progress$.stages(Progress.scala:47)
	at vct.main.stages.Backend.run(Backend.scala:145)
	at vct.main.stages.Backend.run$(Backend.scala:137)
	at vct.main.stages.SilverBackend.run(Backend.scala:184)
	at vct.main.stages.SilverBackend.run(Backend.scala:184)
	at hre.stages.Stages.$anonfun$run$3(Stages.scala:104)
	at hre.stages.Stages.$anonfun$run$3$adapted(Stages.scala:101)
	at scala.collection.IterableOnceOps.foreach(IterableOnce.scala:576)
	at scala.collection.IterableOnceOps.foreach$(IterableOnce.scala:574)
	at scala.collection.AbstractIterable.foreach(Iterable.scala:933)
	at scala.collection.IterableOps$WithFilter.foreach(Iterable.scala:903)
	at hre.stages.Stages.$anonfun$run$1(Stages.scala:101)
	at hre.progress.task.NameSequenceTask.scope(NameSequenceTask.scala:16)
	at hre.progress.Progress$.stages(Progress.scala:47)
	at hre.stages.Stages.run(Stages.scala:98)
	at hre.stages.Stages.run$(Stages.scala:95)
	at hre.stages.StagesPair.run(Stages.scala:145)
	at vct.main.modes.Verify$.verifyWithOptions(Verify.scala:65)
	at vct.main.modes.Verify$.$anonfun$runOptions$3(Verify.scala:100)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.util.Time$.logTime(Time.scala:23)
	at vct.main.modes.Verify$.runOptions(Verify.scala:100)
	at vct.main.Main$.runMode(Main.scala:108)
	at vct.main.Main$.$anonfun$runOptions$3(Main.scala:101)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.middleware.Middleware$.using(Middleware.scala:78)
	at vct.main.Main$.$anonfun$runOptions$2(Main.scala:101)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.io.Watch$.booleanWithWatch(Watch.scala:58)
	at vct.main.Main$.$anonfun$runOptions$1(Main.scala:101)
	at scala.runtime.java8.JFunction0$mcI$sp.apply(JFunction0$mcI$sp.scala:17)
	at hre.middleware.Middleware$.using(Middleware.scala:78)
	at vct.main.Main$.runOptions(Main.scala:96)
	at vct.main.Main$.main(Main.scala:51)
	at vct.main.Main.main(Main.scala)

11:27:25.384 [ERROR] !*!*!*!*!*!*!*!*!*!*!*!
11:27:25.384 [ERROR] ! VerCors has crashed !
11:27:25.384 [ERROR] !*!*!*!*!*!*!*!*!*!*!*!
11:27:25.384 [ERROR] Please report this as a bug here:

```
</details>

--- a/src/col/vct/col/util/StatementToExpression.scala
+++ b/src/col/vct/col/util/StatementToExpression.scala
@@ -81,6 +81,7 @@ case class StatementToExpression[Pre <: Generation, Post <: Generation](
             case _ => None
           }
         x
+      case Scope(_, body) => countAssignments(v, body)
       case Branch(conds) =>
         val assignmentCounts = conds.map(_._2).map(countAssignments(v, _))
           .collect {

--- a/src/main/vct/main/stages/Transformation.scala
+++ b/src/main/vct/main/stages/Transformation.scala
@@ -30,6 +30,7 @@ import vct.result.VerificationError.SystemError
 import vct.rewrite.adt.{EncodeBitVectors, ImportSetCompat}
 import vct.rewrite.{
   CTypeConversions,
+  CollectLocalDeclarations,
   DisambiguateLocation,
   DisambiguatePredicateExpression,
   EncodeAssuming,
@@ -48,6 +49,7 @@ import vct.rewrite.{
   LowerHeapVariables,
   MakeUniqueMethodCopies,
   MonomorphizeClass,
+  PrettifyBlocks,
   SmtlibToProverTypes,
   TypeQualifierCoercion,
   VariableToPointer,

--- a/src/rewrite/vct/rewrite/CollectLocalDeclarations.scala
+++ b/src/rewrite/vct/rewrite/CollectLocalDeclarations.scala
@@ -1,11 +1,8 @@
-package vct.col.rewrite
+package vct.rewrite
 
 import vct.col.ast._
 import vct.col.rewrite.{Generation, Rewriter, RewriterBuilder}
-import vct.col.ast.RewriteHelpers._
 import vct.result.VerificationError.Unreachable
-
-import scala.collection.mutable.ArrayBuffer
 
 case object CollectLocalDeclarations extends RewriterBuilder {
   override def key: String = "collectLocalDecls"
@@ -30,6 +27,6 @@ case class CollectLocalDeclarations[Pre <: Generation]() extends Rewriter[Pre] {
           dispatch(impl)
         }
         Scope[Post](newVars, newImpl)(stat.o)
-      case other => rewriteDefault(other)
+      case _ => super.dispatch(stat)
     }
 }

--- a/src/rewrite/vct/rewrite/EncodeExtract.scala
+++ b/src/rewrite/vct/rewrite/EncodeExtract.scala
@@ -187,7 +187,13 @@ case class EncodeExtract[Pre <: Generation]() extends Rewriter[Pre] {
 
       case extract @ Extract(body, decreases) =>
         body match {
-          // Move scope outside
+          // Move scope outside (expecting no additional statements between the start of the scope and the loop/frame)
+          case scope @ Scope(_, Block(Seq(body))) =>
+            dispatch(
+              scope.copy(body =
+                Extract(body, decreases)(extract.blame)(extract.o)
+              )(scope.o)
+            )
           case scope: Scope[Pre] =>
             dispatch(
               scope.copy(body =

--- a/src/rewrite/vct/rewrite/EncodeExtract.scala
+++ b/src/rewrite/vct/rewrite/EncodeExtract.scala
@@ -187,6 +187,13 @@ case class EncodeExtract[Pre <: Generation]() extends Rewriter[Pre] {
 
       case extract @ Extract(body, decreases) =>
         body match {
+          // Move scope outside
+          case scope: Scope[Pre] =>
+            dispatch(
+              scope.copy(body =
+                Extract(scope.body, decreases)(extract.blame)(extract.o)
+              )(scope.o)
+            )
           case Loop(init, cond, _, blame @ LoopInvariant(invariant, _), _) =>
             val targetValuePair =
               init match {

--- a/src/rewrite/vct/rewrite/PrettifyBlocks.scala
+++ b/src/rewrite/vct/rewrite/PrettifyBlocks.scala
@@ -1,12 +1,7 @@
-package vct.col.rewrite
+package vct.rewrite
 
-import hre.util.ScopedStack
 import vct.col.ast._
-import vct.col.rewrite.{Generation, Rewriter, Rewritten}
-import RewriteHelpers._
-import vct.col.util.AstBuildHelpers._
-
-import scala.collection.mutable.ArrayBuffer
+import vct.col.rewrite.{Generation, Rewriter, RewriterBuilder}
 
 object PrettifyBlocks extends RewriterBuilder {
   override def key: String = "prettifyBlocks"
@@ -15,25 +10,10 @@ object PrettifyBlocks extends RewriterBuilder {
 }
 
 case class PrettifyBlocks[Pre <: Generation]() extends Rewriter[Pre] {
-  def collectVariables(
-      body: Statement[Pre],
-      extra: Seq[Variable[Pre]] = Nil,
-  ): Statement[Post] = {
-    val (variablesHere, newBody) = variables.collect {
-      extra.foreach(dispatch)
-      dispatch(body)
-    }
-
-    Scope(variablesHere, newBody)(body.o)
-  }
-
-  def dispatchFlatly(stat: Statement[Pre]): Seq[Statement[Post]] =
+  private def dispatchFlatly(stat: Statement[Pre]): Seq[Statement[Post]] =
     stat match {
       case Block(statements) => statements.flatMap(dispatchFlatly)
-      case Scope(Nil, body) => dispatchFlatly(body)
-      case Scope(locals, body) if variables.nonEmpty =>
-        locals.foreach(dispatch)
-        dispatchFlatly(body)
+      case Scope(Nil, Block(Nil)) => Nil
       case other => Seq(dispatch(other))
     }
 
@@ -41,25 +21,14 @@ case class PrettifyBlocks[Pre <: Generation]() extends Rewriter[Pre] {
     stat match {
       case Block(statements) =>
         Block(statements.flatMap(dispatchFlatly))(stat.o)
-      case Scope(Nil, body) => dispatch(body)
+      case Scope(Nil, Block(Nil)) => Block[Post](Nil)(stat.o)
       case Scope(locals, body) =>
-        if (variables.nonEmpty) {
+        val (variablesHere, newBody) = variables.collect {
           locals.foreach(dispatch)
           dispatch(body)
-        } else { collectVariables(body, extra = locals) }
-
-      case loop @ Loop(_, _, _, _: IterationContract[Pre], body) =>
-        loop.rewrite(body = collectVariables(body))
-
-      case vec: VecBlock[Pre] =>
-        vec.rewrite(content = collectVariables(vec.content))
-
-      case act: ModelDo[Pre] => act.rewrite(impl = collectVariables(act.impl))
-
-      case pack: WandPackage[Pre] =>
-        pack.rewrite(proof = collectVariables(pack.proof))
-
-      case other => rewriteDefault(other)
+        }
+        Scope(variablesHere, newBody)(body.o)
+      case _ => super.dispatch(stat)
     }
 
   override def dispatch(e: Expr[Pre]): Expr[Post] =
@@ -67,30 +36,6 @@ case class PrettifyBlocks[Pre <: Generation]() extends Rewriter[Pre] {
       case ScopedExpr(locals, body) =>
         locals.foreach(dispatch)
         dispatch(body)
-      case other => rewriteDefault(other)
-    }
-
-  override def dispatch(parRegion: ParRegion[Pre]): ParRegion[Rewritten[Pre]] =
-    parRegion match {
-      case block: ParBlock[Pre] =>
-        block.rewrite(content = collectVariables(block.content))
-      case other => rewriteDefault(other)
-    }
-
-  override def dispatch(decl: Declaration[Pre]): Unit =
-    decl match {
-      case method: RunMethod[Pre] =>
-        classDeclarations.succeed(
-          method,
-          method.rewrite(body = method.body.map(collectVariables(_))),
-        )
-      case method: AbstractMethod[Pre] =>
-        allScopes.anySucceedOnly(
-          method,
-          allScopes.anyDeclare(
-            method.rewrite(body = method.body.map(collectVariables(_)))
-          ),
-        )
-      case other => rewriteDefault(other)
+      case _ => super.dispatch(e)
     }
 }

--- a/src/rewrite/vct/rewrite/PrettifyBlocks.scala
+++ b/src/rewrite/vct/rewrite/PrettifyBlocks.scala
@@ -2,6 +2,7 @@ package vct.rewrite
 
 import vct.col.ast._
 import vct.col.rewrite.{Generation, Rewriter, RewriterBuilder}
+import vct.result.VerificationError.Unreachable
 
 object PrettifyBlocks extends RewriterBuilder {
   override def key: String = "prettifyBlocks"
@@ -33,9 +34,13 @@ case class PrettifyBlocks[Pre <: Generation]() extends Rewriter[Pre] {
 
   override def dispatch(e: Expr[Pre]): Expr[Post] =
     e match {
-      case ScopedExpr(locals, body) =>
+      case ScopedExpr(locals, body) if variables.nonEmpty =>
         locals.foreach(dispatch)
         dispatch(body)
+      case ScopedExpr(_, _) =>
+        throw Unreachable(
+          s"Got to this scoped expression (`$e`) but there is no outer scope. Every function/method body should start with a Scope node"
+        )
       case _ => super.dispatch(e)
     }
 }

--- a/src/rewrite/vct/rewrite/lang/LangCPPToCol.scala
+++ b/src/rewrite/vct/rewrite/lang/LangCPPToCol.scala
@@ -1475,7 +1475,10 @@ case class LangCPPToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
     val kernelRunner =
       new RunMethod[Post](
         body = Some(
-          ParStatement[Post](kernelParBlock)(kernelDeclaration.body.o)
+          Scope(
+            Nil,
+            ParStatement[Post](kernelParBlock)(kernelDeclaration.body.o),
+          )(kernelDeclaration.body.o)
         ),
         contract = kernelRunnerContract,
       )(KernelLambdaRunMethodBlame(kernelDeclaration))(commandGroup.o)

--- a/src/rewrite/vct/rewrite/lang/LangJavaToCol.scala
+++ b/src/rewrite/vct/rewrite/lang/LangJavaToCol.scala
@@ -297,11 +297,14 @@ case class LangJavaToCol[Pre <: Generation](rw: LangSpecificToCol[Pre])
                 outArgs = Nil,
                 typeArgs = Nil,
                 body = Some(rw.currentThis.having(`this`) {
-                  Block(Seq(
-                    fieldInit(`this`),
-                    sharedInit(`this`),
-                    rw.dispatch(cons.body),
-                  ))
+                  Scope(
+                    Nil,
+                    Block(Seq(
+                      fieldInit(`this`),
+                      sharedInit(`this`),
+                      rw.dispatch(cons.body),
+                    )),
+                  )
                 }),
                 contract =
                   rw.currentThis.having(`this`) {

--- a/src/rewrite/vct/rewrite/veymont/generation/GenerateImplementation.scala
+++ b/src/rewrite/vct/rewrite/veymont/generation/GenerateImplementation.scala
@@ -369,6 +369,8 @@ case class GenerateImplementation[Pre <: Generation]()
         exhale.rewrite(res = projectExpr(exhale.res))
       // Rewrite blocks transparently
       case block: Block[Pre] => block.rewriteDefault()
+      // Similarly, for scopes
+      case scope: Scope[Pre] => scope.rewriteDefault()
       // Plain assigns were warned about in LangVeyMontToCol.
       // We just keep them in the program here for debugging purposes.
       case assign: Assign[Pre] => assign.rewriteDefault()

--- a/src/rewrite/vct/rewrite/veymont/generation/SpecializeEndpointClasses.scala
+++ b/src/rewrite/vct/rewrite/veymont/generation/SpecializeEndpointClasses.scala
@@ -2,8 +2,8 @@ package vct.rewrite.veymont.generation
 
 import com.typesafe.scalalogging.LazyLogging
 import vct.col.ast.{
-  ChorRun,
   ByReferenceClass,
+  ChorRun,
   Choreography,
   Class,
   Communicate,
@@ -21,6 +21,7 @@ import vct.col.ast.{
   LoopInvariant,
   Program,
   Receiver,
+  Scope,
   Sender,
   ThisObject,
   UnitAccountedPredicate,
@@ -100,11 +101,14 @@ case class SpecializeEndpointClasses[Pre <: Generation]()
                   )) === implArg.get)
               ),
             ),
-            body = Some(assignField[Post](
-              ThisObject(classOfEndpoint.ref(endpoint)),
-              implField.ref,
-              implArg.get,
-              PanicBlame("Cannot fail"),
+            body = Some(Scope(
+              Nil,
+              assignField[Post](
+                ThisObject(classOfEndpoint.ref(endpoint)),
+                implField.ref,
+                implArg.get,
+                PanicBlame("Cannot fail"),
+              ),
             )),
             outArgs = Seq(),
             typeArgs = Seq(),

--- a/src/viper/viper/api/backend/SilverBackend.scala
+++ b/src/viper/viper/api/backend/SilverBackend.scala
@@ -4,7 +4,8 @@ import com.typesafe.scalalogging.LazyLogging
 import hre.io.RWFile
 import hre.progress.Progress
 import vct.col.ast.Node
-import vct.col.origin.AccountedDirection
+import vct.col.origin.{AccountedDirection, FailLeft, FailRight}
+import vct.col.ref.Ref
 import vct.col.{ast => col, origin => blame}
 import vct.result.VerificationError.{SystemError, TimeOut}
 import viper.api.SilverTreeCompare
@@ -33,7 +34,6 @@ import viper.silver.{ast => silver}
 
 import java.nio.file.{Files, Path}
 import scala.reflect.ClassTag
-import scala.util.matching.Regex
 import scala.util.{Try, Using}
 
 trait SilverBackend
@@ -202,8 +202,19 @@ trait SilverBackend
             ))
           case PreconditionInAppFalse(node, reason, _) =>
             val invocation = get[col.FunctionInvocation[_]](node)
+            val offendingArgs = node.args.zipWithIndex.collect {
+              case (a, i)
+                  if System.identityHashCode(a) ==
+                    System.identityHashCode(reason.offendingNode) =>
+                i
+            }
+            // Using collectFirst here, not sure how you would realistically get more than one here, but I guess if there's multiple they're probably identical
+            val offendingPath = offendingArgs.map(invocation.ref.decl.args(_))
+              .collectFirst(Function.unlift(
+                argToPath(invocation.ref.decl.contract.requires, _)
+              )).getOrElse(path(reason.offendingNode))
             invocation.blame.blame(blame.PreconditionFailed(
-              path(reason.offendingNode),
+              offendingPath,
               getFailure(reason),
               invocation,
             ))
@@ -483,5 +494,24 @@ trait SilverBackend
         throw NotSupported(
           s"Viper returned an error reason that VerCors does not recognize: $other"
         )
+    }
+
+  private def argInExpr(e: col.Expr[_], arg: col.Variable[_]): Boolean =
+    e match {
+      case col.Local(Ref(v)) => v == arg
+      case col.And(l, r) => argInExpr(l, arg) || argInExpr(r, arg)
+      case col.Star(l, r) => argInExpr(l, arg) || argInExpr(r, arg)
+      case _ => false
+    }
+  private def argToPath(
+      p: col.AccountedPredicate[_],
+      arg: col.Variable[_],
+  ): Option[Seq[AccountedDirection]] =
+    p match {
+      case col.UnitAccountedPredicate(e) if argInExpr(e, arg) => Some(Nil)
+      case col.UnitAccountedPredicate(_) => None
+      case col.SplitAccountedPredicate(left, right) =>
+        argToPath(left, arg).map(FailLeft +: _)
+          .orElse(argToPath(right, arg).map(FailRight +: _))
     }
 }

--- a/src/viper/viper/api/backend/SilverBackend.scala
+++ b/src/viper/viper/api/backend/SilverBackend.scala
@@ -244,6 +244,7 @@ trait SilverBackend
                 case reasons.MagicWandChunkNotFound(n) => n
                 case reasons.AssertionFalse(n) => n
                 case reasons.NegativePermission(n) => n
+                case reasons.QPAssertionNotInjective(n) => n
               }
             val (bl, assert) = info(offNode).asserting.map(n => (n.blame, n))
               .getOrElse[(blame.Blame[blame.AssertFailed], Node[_])](

--- a/test/viper/viper/api/TestSilicon.scala
+++ b/test/viper/viper/api/TestSilicon.scala
@@ -25,7 +25,7 @@ class TestSilicon extends VerifySpec[(silver.Program, Map[Int, Node[_]])](Silico
   )
 
   vercors should "verify a dereference with sufficient permission" in procedure(
-    args=Seq(r), requires=Perm(SilverFieldLocation(r.get, int.ref), WritePerm()),
+    args=Seq(r), requires=Perm(SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm())),
     body=Scope(Seq(i), i <~ r.get~>int)
   )
 
@@ -34,7 +34,7 @@ class TestSilicon extends VerifySpec[(silver.Program, Map[Int, Node[_]])](Silico
   )
 
   vercors should "assign a field with sufficient permission" in procedure(
-    args=Seq(r), requires=Perm(SilverFieldLocation(r.get, int.ref), WritePerm()),
+    args=Seq(r), requires=Perm(SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm())),
     body=r.get~>int <~ const(0)
   )
 
@@ -53,7 +53,7 @@ class TestSilicon extends VerifySpec[(silver.Program, Map[Int, Node[_]])](Silico
     body=Assert[G](
       Starall(Seq(i), Seq(),
         Implies(i.get >= const(0) && i.get < SilverSeqSize(rs.get),
-          Perm(SilverFieldLocation[G]((rs.get @@ i.get), int.ref), WritePerm())))(ExpectError()))(noErrors)
+          Perm(SilverFieldLocation[G]((rs.get @@ i.get), int.ref), SilverIntToRat(WritePerm()))))(ExpectError()))(noErrors)
   )
 
   val p = new Variable[G](TRational())
@@ -63,18 +63,18 @@ class TestSilicon extends VerifySpec[(silver.Program, Map[Int, Node[_]])](Silico
   )
 
   vercors should "report insufficient permission to exhale when asserting too much permission" in procedure(
-    args=Seq(r), body=Block[G](Seq(Assert[G](Perm(SilverFieldLocation(r.get, int.ref), WritePerm()))(ExpectError())))
+    args=Seq(r), body=Block[G](Seq(Assert[G](Perm(SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm())))(ExpectError())))
   )
 
   vercors should "verify a valid exhale of permission" in procedure(
-    args=Seq(r), requires=Perm(SilverFieldLocation[G](r.get, int.ref), WritePerm()),
-    body=Block[G](Seq(Exhale[G](Perm(SilverFieldLocation(r.get, int.ref), WritePerm()))(noErrors)))
+    args=Seq(r), requires=Perm(SilverFieldLocation[G](r.get, int.ref), SilverIntToRat(WritePerm())),
+    body=Block[G](Seq(Exhale[G](Perm(SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm())))(noErrors)))
   )
 
   vercors should "report insufficient permission to exhale when exhaling too much permission" in procedure(
-    args=Seq(r), body=Exhale[G](Perm(SilverFieldLocation(r.get, int.ref), WritePerm()))(ExpectError())
+    args=Seq(r), body=Exhale[G](Perm(SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm())))(ExpectError())
   )
 
-  val validPred = new Predicate(Seq(), Some(Perm[G](SilverFieldLocation(r.get, int.ref), WritePerm())))
+  val validPred = new Predicate(Seq(), Some(Perm[G](SilverFieldLocation(r.get, int.ref), SilverIntToRat(WritePerm()))))
   val invalidPred = new Predicate(Seq(), Some(Eq(r.get~>int, const(5))))
 }

